src/VecSim/utils/merge_results.h

```diff
@@ -0,0 +1,79 @@
+/*
+ *Copyright Redis Ltd. 2021 - present
+ *Licensed under your choice of the Redis Source Available License 2.0 (RSALv2) or
+ *the Server Side Public License v1 (SSPLv1).
+ */
+
+#pragma once
+
+#include "VecSim/query_result_struct.h"
+#include <VecSim/utils/vec_utils.h>
+#include "arr_cpp.h"
+#include <unordered_set>
+
+static int cmpVecSimQueryResultByScoreThenId(const VecSimQueryResult *res1,
+                                             const VecSimQueryResult *res2) {
+    return (VecSimQueryResult_GetScore(res1) != VecSimQueryResult_GetScore(res2))
+               ? cmpVecSimQueryResultByScore(res1, res2)
+               : cmpVecSimQueryResultById(res1, res2);
+}
+
+// Assumes that the arrays are sorted by score firstly and by id secondarily.
+template <bool withSet>
+VecSimQueryResult_List merge_results(VecSimQueryResult_List first, VecSimQueryResult_List second,
+                                     size_t limit) {
+
+    VecSimQueryResult *results = array_new<VecSimQueryResult>(limit);
+    VecSimQueryResult_List mergedResults{.results = results, .code = VecSim_QueryResult_OK};
+    VecSimQueryResult *cur_first = first.results;
+    VecSimQueryResult *cur_second = second.results;
+    const auto first_end = cur_first + VecSimQueryResult_Len(first);
+    const auto second_end = cur_second + VecSimQueryResult_Len(second);
+
+    // Will hold the ids of the results we've already added to the merged results.
+    // Will be used only if withSet is true.
+    std::unordered_set<size_t> ids;
```

Comment:
it might... but the set/number of lit bits will be `k` at most so I can't tell what will be better. we might need to revisit this.
Maybe we should look for a sparse bitmap implementation?

Reply:
All right, let's keep it like this